wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/convert/DefaultEditorProvisioner.java

```diff
@@ -99,20 +104,97 @@ public void apply(DevfileImpl devfile, FileContentProvider contentProvider)
+   * @throws DevfileException
+   */
   private void provisionDefaultPlugins(
       List<ComponentImpl> components, FileContentProvider contentProvider) throws DevfileException {
     Map<String, String> missingPluginsIdToRef = new HashMap<>(defaultPluginsToRefs);
+    removeAlreadyAddedPlugins(components, contentProvider, missingPluginsIdToRef);
+    addMissingPlugins(components, contentProvider, missingPluginsIdToRef);
+  }
 
+  /**
+   * Checks if any of the Devfile's components are also in the list of missing default plugins, and
+   * removes them.
+   *
+   * @param components - The list of Devfile components
+   * @param contentProvider - The content provider to retrieve YAML
+   * @param missingPluginsIdToRef - The list of default plugins that are not currently in the list
+   *     of Devfile components
+   */
+  private void removeAlreadyAddedPlugins(
+      List<ComponentImpl> components,
+      FileContentProvider contentProvider,
+      Map<String, String> missingPluginsIdToRef)
+      throws DevfileException {
     for (ComponentImpl t : components) {
       if (PLUGIN_COMPONENT_TYPE.equals(t.getType())) {
         String pluginPublisherAndName = getPluginPublisherAndName(t, contentProvider);
         missingPluginsIdToRef.remove(pluginPublisherAndName);
       }
     }
+  }
 
-    missingPluginsIdToRef
-        .values()
-        .forEach(pluginRef -> components.add(new ComponentImpl(PLUGIN_COMPONENT_TYPE, pluginRef)));
+  /**
+   * Tries to add default plugins to the Devfile components. Each plugin is initially parsed by
+   * plugin ref. If the plugin does not have a reference, it is added to the component list, and its
+   * plugin ID will be used to resolve it. If it has a reference, the Plugin is evaluated, so that
+   * its meta.yaml can be retrieved. From the meta.yaml, the new Component's ID and reference are
+   * properly set, and the Component is added to the list.
+   *
+   * @param components - The list of Devfile components
+   * @param contentProvider - The content provider to retrieve YAML
+   * @param missingPluginsIdToRef - The list of default plugins that are not currently in the list
+   *     of devfile components
+   * @throws DevfileException
+   */
+  private void addMissingPlugins(
+      List<ComponentImpl> components,
+      FileContentProvider contentProvider,
+      Map<String, String> missingPluginsIdToRef)
+      throws DevfileException {
+    for (String pluginRef : missingPluginsIdToRef.values()) {
+      ComponentImpl component;
+      try {
+        ExtendedPluginFQN fqn = pluginFQNParser.parsePluginFQN(pluginRef);
+        if (!isNullOrEmpty(fqn.getId())) {
+          component = new ComponentImpl(PLUGIN_COMPONENT_TYPE, pluginRef);
+        } else {
+          component = createReferencePluginComponent(pluginRef, contentProvider);
+        }
+        components.add(component);
+      } catch (InfrastructureException e) {
+        throw new DevfileException(e.getMessage(), e);
```

Comment:
What is reason to wrap `InfrastructureException` into `DevfileException` ?

Reply:
I changed the methods to throw the underlying `InfrastructureExceptions`.  `provisionDefaultPlugins` throws a `DevfileException`, so the try/catch there wraps the `InfrastructureException` into a `DevfileException` so I don't break that contract.